File: app/src/main/java/github/activity/ui/ActivityColor.java
package github.activity.ui;

import android.graphics.Color;

/**
* Created by pavel on 28/03/15.
*/
public class ActivityColor {

	int alpha;
	float h, s, v;

	public ActivityColor(int color) {
		alpha = Color.alpha(color);
		float[] hsv = new float[3];
		Color.colorToHSV(color, hsv);
		h = hsv[0];
		s = hsv[1];
		v = hsv[2];
	}

	public static int calculateColor(ActivityColor minColor, ActivityColor maxColor, int maxActivity, int currentActivity) {

		float coeff = (float) currentActivity / maxActivity;

		int a = (int) (minColor.getAlpha() + (maxColor.getAlpha() - minColor.getAlpha()) * coeff);
		float h = minColor.getH() + (maxColor.getH() - minColor.getH()) * coeff;
		float s = minColor.getS() + (maxColor.getS() - minColor.getS()) * coeff;
		float v = minColor.getV() + (maxColor.getV() - minColor.getV()) * coeff;

		int color = Color.HSVToColor(new float[]{h, s, v});
		color = Color.argb(a, Color.red(color), Color.green(color), Color.blue(color));
		return color;
	}

	public int getAlpha() {
		return alpha;
	}

	public float getH() {
		return h;
	}

	public float getS() {
		return s;
	}

	public float getV() {
		return v;
	}

}

File: app/src/main/java/github/activity/ui/ActivityColor.kt
package github.activity.ui

import android.graphics.Color

data class ActivityColor(
	val alpha: Int,
	val h: Float,
	val s: Float,
	val v: Float
) {

	companion object {

		fun fromIntValue(color: Int): ActivityColor {
			val hsv = FloatArray(3)
			Color.colorToHSV(color, hsv)
			val alpha = Color.alpha(color)
			val h = hsv[0]
			val s = hsv[1]
			val v = hsv[2]
			return ActivityColor(alpha, h, s, v)
		}

		fun calculateColor(
			minColor: ActivityColor,
			maxColor: ActivityColor,
			maxActivity: Int,
			currentActivity: Int
		): Int {

            val coeff = currentActivity.toFloat() / maxActivity

            val a = (minColor.alpha + (maxColor.alpha - minColor.alpha) * coeff).toInt()
            val h = minColor.h + (maxColor.h - minColor.h) * coeff
            val s = minColor.s + (maxColor.s - minColor.s) * coeff
            val v = minColor.v + (maxColor.v - minColor.v) * coeff

            var color = Color.HSVToColor(floatArrayOf(h, s, v))
            color = Color.argb(
				a,
				Color.red(color),
				Color.green(color),
				Color.blue(color)
			)
            return color
        }

    }

}
